### ISSUE TITLE
Maximum call stack size exceeded

### ISSUE THREAD

**Issue by fazibear** [elixirscript/elixirscript#221]
There is no way to loop something. Example code generate exception.

```elixir
defmodule MyApp do
  def main() do
    a = 1
    b = 2
    loop(a,b)
  end

  def loop(a,b) do
    loop(a,b)
  end
end
```

```
Uncaught RangeError: Maximum call stack size exceeded
    at fillInOptionalValues (MyApp.js:1600)
    at MyApp.js:1576
    at MyApp.js:2769
    at MyApp.js:1579
    at MyApp.js:2769
    at MyApp.js:1579
    at MyApp.js:2769
    at MyApp.js:1579
    at MyApp.js:2769
    at MyApp.js:1579
```

**Comment by Qqwy** [comment]
As @bryanjos and I were talking about today on Slack, to support this, ElixirScript needs Tail Call elimination (AKA Tail Recursion) support.


On the JS/ESTree side of things, this basically means that before a function definition is compiled to JS, its ESTree needs to be altered from:

```elixir
%ESTree.ReturnStatement{argument: old_code}
```
to something akin:
```elixir

%ESTree.ReturnStatement{argument: 
  %ESTree.FunctionExpression{body: 
    %ESTree.BlockStatement{body: [old_expression_here] }}
}
````

That is, the resulting JS will instead of looking something like:

```js
function(){
  foo();
  return bar();
}
```
look like:
```js
function(){
  foo();
  return function(){ bar() };
}
```
Now, the outernmost call that starts the script should not simply call the entry point, but be changed to a trampoline function like this:

```js
function(){
  while(typeof(f) === "function"){
    f = f();
  }
}
```

Furthermore, if this trampoline is altered to something akin:

```js
function*(){
  while(typeof(f) === "function"){
    f = f();
    yield;
  }
}
```
then this is a generator that might be scheduled.

**Comment by bryanjos** [comment]
@Qqwy thanks for the explanation. I'm still trying to absorb it.

When you say "Now, the outernmost call that starts the script should not simply call the entry point", do you mean like instead of calling `bar()` at the top level, it should be wrapped in something that calls it instead?

Also should the code look at the return, check to see if the function being called last is the same name as the current function and wrap the function in something to handle the tail call?

I've seen some other links that look like they are along the same paths:
https://gist.github.com/Integralist/11246383
https://github.com/Gozala/js-tail-call

**Comment by Qqwy** [comment]
> When you say "Now, the outernmost call that starts the script should not simply call the entry point", do you  mean like instead of calling bar() at the top level, it should be wrapped in something that calls it instead?

Yep. instead of `bar(x,y,z)` you should call `trampoline(function(){bar(x,y,z)})`.

Also, what I did not realize before is that besides doing this at the outernmost level, if you want to use the result of an _inner_ function call, you'll also need to wrap that inner call in a trampoline, to ensure that those recursive calls are all followed.

**Comment by Qqwy** [comment]
A better example code to test if TCO works is the following (since the example in the first post will, instead of generating an error, 'just hang' infinitely in a tail-call optimized language):

```
defmodule Fibonacchi do

  def fib(num) when num <= 0 do
    0
  end

  def fib(1) do
    1
  end

  def fib(n) do
    fib(n - 1) + fib(n - 2)
  end
end

Fibonacchi.fib(100)
```

**Comment by bryanjos** [comment]
@Qqwy thanks, that helps. So would I then call `fact` like so:

```js
fact(10000)
```

or would I have to call it like this:

```js
trampoline(function(){ return fact(10000) });
```

And looks like the only concern is when a function is called in the return correct?

**Comment by bryanjos** [comment]
Going to try to give this a shot tonight

**Comment by bryanjos** [comment]
@Qqwy I have this code in Elixirscript

```elixir
defmodule Factorial do

  def fact(num) do
    do_fact(num, 1)
  end

  defp do_fact(0, acc), do: acc
  defp do_fact(num, acc), do: do_fact(num - 1, acc * num)
end

defmodule App do
  @on_js_load :main

  def main() do
    Factorial.fact(10000)
  end
end
```

That produces this javascript


```js
//Factorial.ex
import Elixir from '../elixir/Elixir.Bootstrap';
import Elixir$ElixirScript$Kernel from '../elixir/Elixir.ElixirScript.Kernel';

const do_fact = Elixir.Core.Patterns.defmatch(
  Elixir.Core.Patterns.clause([0, Elixir.Core.Patterns.variable()], 
    function(acc) {
      return acc;
    }
  ), 
  Elixir.Core.Patterns.clause([Elixir.Core.Patterns.variable(), Elixir.Core.Patterns.variable()],      
    function(num, acc) {
      return Elixir.Core.Functions.trampoline(function() {
        return do_fact(num - 1, acc * num);
    });
  })
);

const fact = Elixir.Core.Patterns.defmatch(Elixir.Core.Patterns.clause([Elixir.Core.Patterns.variable()],   
  function(num) {
    return Elixir.Core.Functions.trampoline(function() {
        return do_fact(num, 1);
    });
}));

export default {
    fact
};


//App.ex
import Elixir from '../elixir/Elixir.Bootstrap';
import Elixir$ElixirScript$Kernel from '../elixir/Elixir.ElixirScript.Kernel';
import Elixir$Factorial from '../app/Elixir.Factorial';

const main = Elixir.Core.Patterns.defmatch(Elixir.Core.Patterns.clause([], 
  function() {
    return Elixir.Core.Functions.trampoline(function() {
        return Elixir$Factorial.fact(10000);
    });
}));

main();

export default {
    main
};

```

Notice how functions are translated in elixirscript (to allow pattern matching). Also notice the trampoline calls whenever a function is called as the last line.

I'm still getting stack errors

```
Elixir.Bootstrap.js:2105 Uncaught RangeError: Maximum call stack size exceeded
    at Object.trampoline (Elixir.Bootstrap.js:2105)
    at eval (Elixir.Factorial.js:10)
    at result (Elixir.Bootstrap.js:1579)
    at eval (Elixir.Factorial.js:11)
    at Object.trampoline (Elixir.Bootstrap.js:2106)
    at eval (Elixir.Factorial.js:10)
    at result (Elixir.Bootstrap.js:1579)
    at eval (Elixir.Factorial.js:11)
    at Object.trampoline (Elixir.Bootstrap.js:2106)
    at eval (Elixir.Factorial.js:10)
```


I wonder though if maybe it has more to do with the way functions are translated. Maybe I should have the trampoline call something else?

Like maybe here? https://github.com/bryanjos/tailored/blob/master/src/tailored/defmatch.js#L46

**Comment by bryanjos** [comment]
Ok I think I solved it! I have to make a change to the tailored lib. That's the only place needed.

Here is what I ended up with:

```js
function trampoline(fn){  
    return function(){
        var res = fn.apply(this, arguments);
        while(res instanceof Function){
            res = res();
        }
        return res;
    }
}

export function defmatch(...clauses) {
  return trampoline(function(...args) {
    let funcToCall = null;
    let params = null;
    for (let processedClause of clauses) {
      let result = [];
      args = fillInOptionalValues(args, processedClause.arity, processedClause.optionals);

      if (processedClause.pattern(args, result) && processedClause.guard.apply(this, result)) {
        funcToCall = processedClause.fn;
        params = result;
        break;
      }
    }

    if (!funcToCall) {
      console.error('No match for:', args);
      throw new MatchError(args);
    }

    return funcToCall.bind(this, ...params);
  });
}
```

It makes sense now, but hand to make sure it returns a function. Then the trampolining could work.

I can make an update to everything tomorrow

**Comment by Qqwy** [comment]
Awesome work!

**Comment by bryanjos** [comment]
Fixed in 0.24.0

**Comment by fazibear** [comment]
Did you test it? It doesn't work for me?

**Comment by bryanjos** [comment]
I tested it with the fibonacci example not the example from the original. I can test it out with that one later today to make sure.

**Comment by Qqwy** [comment]
I expect that the original example will still not work since it will hang in an infinite loop without anything happening. (It might crash your browser tab?).

**Comment by fazibear** [comment]
No. I get the same exception with `Maximum call stack size exceeded`

**Comment by bryanjos** [comment]
Yeah I'm getting the same as well. The change was in [this library](https://github.com/bryanjos/tailored) and I had a test using fibonacci [here](https://github.com/bryanjos/tailored/blob/master/test/tailcall.spec.js). It only has 10 but I tested it with much higher numbers. I made a test locally in that repo with what is represented here and sure enough I got a stack exception. 

I haven't had a chance to revisit it just yet, but that's where the change probably needs to take place
